include/kernel.h

```diff
@@ -3959,6 +3976,219 @@ static inline char *K_THREAD_STACK_BUFFER(k_thread_stack_t sym)
 
 #endif /* _ARCH_DECLARE_STACK */
 
+/**
+ * @defgroup mem_domain_apis Memory domain APIs
+ * @ingroup kernel_apis
+ * @{
+ */
+
+/** @def MEM_PARTITION_ENTRY
+ *  @brief Used to declare a memory partition entry
+ */
+#define MEM_PARTITION_ENTRY(_start, _size, _attr) \
+	{\
+		.start = _start, \
+		.size = _size, \
+		.attr = _attr, \
+	}
+
+/** @def K_MEM_PARTITION_DEFINE
+ *  @brief Used to declare a memory partition
+ */
+#ifdef _ARCH_MEM_PARTITION_ALIGN_CHECK
+#define K_MEM_PARTITION_DEFINE(name, start, size, attr) \
+	_ARCH_MEM_PARTITION_ALIGN_CHECK(start, size); \
+	struct k_mem_partition name = \
+		MEM_PARTITION_ENTRY((u32_t)start, size, attr)
+#else
+#define K_MEM_PARTITION_DEFINE(name, start, size, attr) \
+	struct k_mem_partition name = \
+		MEM_PARTITION_ENTRY((u32_t)start, size, attr)
+#endif /* _ARCH_MEM_PARTITION_ALIGN_CHECK */
+
+/* memory partition arch/soc independent attribute */
+/** @def K_MEM_PARTITION_P_NA_U_NA
+ *  @brief Denote partition is "Privileged No Access, Unprivileged No Access"
+ */
+#define K_MEM_PARTITION_P_NA_U_NA				0
```

Comment:
> But can you please tell me why would you need a memory allocated to you to be set to P_NA. Shouldn't the kernel have the access to that memory?? I maybe be missing some usecase here.

I just want to give the maximum flexibility to developers, so I list all possible settings of the MPU.
Maybe tihs could be used for debugging purpose to trap the buffer overflow issue.